libraries/standard/mqtt/src/iot_mqtt_serialize.c

```diff
@@ -1567,134 +1647,83 @@ IotMqttError_t _IotMqtt_DeserializePublish( _mqttPacket_t * pPublish )
     /* The flags are the lower 4 bits of the first byte in PUBLISH. */
     publishFlags = pPublish->type;
 
-    /* Parse the Retain bit. */
-    pOutput->retain = UINT8_CHECK_BIT( publishFlags, MQTT_PUBLISH_FLAG_RETAIN );
-
-    IotLog( IOT_LOG_DEBUG,
-            &_logHideAll,
-            "Retain bit is %d.", pOutput->retain );
-
-    /* Check for QoS 2. */
-    if( UINT8_CHECK_BIT( publishFlags, MQTT_PUBLISH_FLAG_QOS2 ) == true )
-    {
-        /* PUBLISH packet is invalid if both QoS 1 and QoS 2 bits are set. */
-        if( UINT8_CHECK_BIT( publishFlags, MQTT_PUBLISH_FLAG_QOS1 ) == true )
-        {
-            IotLog( IOT_LOG_DEBUG,
-                    &_logHideAll,
-                    "Bad QoS: 3." );
+    status = _processIncomingPublishFlags( publishFlags, pOutput );
 
-            status = IOT_MQTT_BAD_RESPONSE;
-            goto cleanup;
-        }
-
-        pOutput->qos = IOT_MQTT_QOS_2;
-    }
-    /* Check for QoS 1. */
-    else if( UINT8_CHECK_BIT( publishFlags, MQTT_PUBLISH_FLAG_QOS1 ) == true )
-    {
-        pOutput->qos = IOT_MQTT_QOS_1;
-    }
-    /* If the PUBLISH isn't QoS 1 or 2, then it's QoS 0. */
-    else
+    if( status == IOT_MQTT_SUCCESS )
     {
-        pOutput->qos = IOT_MQTT_QOS_0;
+        /* Sanity checks for "Remaining length". A QoS 0 PUBLISH  must have a remaining
+         * length of at least 3 to accommodate topic name length (2 bytes) and topic
+         * name (at least 1 byte). A QoS 1 or 2 PUBLISH must have a remaining length of
+         * at least 5 for the packet identifier in addition to the topic name length and
+         * topic name. */
+        status = _checkRemainingLength( pPublish, pOutput->qos, MQTT_MIN_PUBLISH_REMAINING_LENGTH_QOS0 );
     }
 
-    IotLog( IOT_LOG_DEBUG,
-            &_logHideAll,
-            "QoS is %d.", pOutput->qos );
-
-    /* Parse the DUP bit. */
-    if( UINT8_CHECK_BIT( publishFlags, MQTT_PUBLISH_FLAG_DUP ) == true )
+    if( status == IOT_MQTT_SUCCESS )
     {
-        IotLog( IOT_LOG_DEBUG,
-                &_logHideAll,
-                "DUP is 1." );
+        /* Extract the topic name starting from the first byte of the variable header.
+         * The topic name string starts at byte 3 in the variable header. */
+        pOutput->topicNameLength = UINT16_DECODE( pVariableHeader );
+
+        /* Sanity checks for topic name length and "Remaining length". The remaining
+         * length must be at least as large as the variable length header. */
+        status = _checkRemainingLength( pPublish,
+                                        pOutput->qos,
+                                        pOutput->topicNameLength + sizeof( uint16_t ) );
```

Comment:
I don't see any way in which we communicate to the user that the buffer is incomplete.  Seems like that would make this hard to use with a streams model.  Am I missing something?